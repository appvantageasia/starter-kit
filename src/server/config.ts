import chalk from 'chalk';
import { Region } from 'minio';
import { SMTPSettings } from './emails';
import { getString, getBoolean, getInteger, getNumber } from './env';

const prefix = 'APP';

const getPrefix = key => `${prefix}_${key}`;

export type RuntimeConfig = {
    version: string;

    port: number;

    publicPath: string;

    db: {
        uri: string;
        name: string;
        pool: number;
    };

    redis: {
        uri: string;
    };

    session: {
        secret: string;
        lifetime: string;
    };

    i18n: {
        locales: string[];
        defaultLocale: string;
    };

    smtp: {
        transporter: SMTPSettings;
        sender: string;
    };

    storage: {
        provider: {
            endPoint: string;
            accessKey: string;
            secretKey: string;
            useSSL?: boolean;
            port?: number;
            region: Region;
        };
        bucket: string;
    };

    sentry: {
        dsn?: string;
        release?: string;
        environment?: string;
        tracing: boolean;
        tracesSampleRate: number;
    };

    html2pdf: {
        endpoint: string;
    };
};

const getSmtpSettings = (): RuntimeConfig['smtp']['transporter'] => {
    const base = {
        host: getString(getPrefix('SMTP_HOST'), 'localhost'),
        port: getInteger(getPrefix('SMTP_PORT'), 465),
        secure: getBoolean(getPrefix('SMTP_SECURE'), false),
    };

    const user = getString(getPrefix('SMTP_USER'));

    if (!user) {
        return base;
    }

    return {
        ...base,
        auth: {
            user,
            pass: getString(getPrefix('SMTP_PASSWORD')),
        },
    };
};

const version = getString('VERSION', '0.0.0-development');

const config: RuntimeConfig = {
    version,

    port: getNumber(getPrefix('PORT'), 3000),
    publicPath: getString(getPrefix('PUBLIC_PATH'), '/public/'),

    // internationalization
    i18n: {
        locales: ['en'],
        defaultLocale: 'en',
    },

    // server runtime
    db: {
        uri: getString(getPrefix('DB_URI'), 'mongodb+srv://localhost:27017'),
        name: getString(getPrefix('DB_NAME'), 'app'),
        pool: getInteger(getPrefix('DB_POOL'), 10),
    },

    redis: {
        uri: getString(getPrefix('REDIS_URI'), 'redis://127.0.0.1:6379'),
    },

    session: {
        secret: getString(getPrefix('SESSION_SECRET')),
        lifetime: getString(getPrefix('SESSION_LIFETIME'), '1h'),
    },

    // default/system SMTP settings
    smtp: {
        transporter: getSmtpSettings(),
        sender: getString(getPrefix('SMTP_FROM'), 'noreply@appvantage.co'),
    },

    storage: {
        provider: {
            endPoint: getString(getPrefix('STORAGE_ENDPOINT')),
            accessKey: getString(getPrefix('STORAGE_ACCESS_KEY')),
            secretKey: getString(getPrefix('STORAGE_SECRET_KEY')),
            useSSL: getBoolean(getPrefix('STORAGE_SSL'), true),
            port: getInteger(getPrefix('STORAGE_PORT')),
            region: getString(getPrefix('STORAGE_REGION'), 'ap-southeast-1'),
        },
        bucket: getString(getPrefix('STORAGE_BUCKET'), 'app'),
    },

    sentry: {
        dsn: getString(getPrefix('SENTRY_DSN')),
        release: getString(getPrefix('SENTRY_RELEASE')),
        environment: getString(getPrefix('SENTRY_ENVIRONMENT')),
        tracing: getBoolean(getPrefix('SENTRY_TRACING'), true),
        tracesSampleRate: getNumber(getPrefix('SENTRY_TRACES_SAMPLE_RATE'), 1.0),
    },

    html2pdf: {
        endpoint: getString(getPrefix('HTML2PDF_ENDPOINT')),
    },
};

const exitOnError = (error: string): void => {
    console.error(error);
    process.exit(1);
};

export const runValidityChecks = () => {
    if (!config.session.secret) {
        exitOnError(chalk.red('APP_SESSION_SECRET is missing in environment variables'));
    }
};

export default config;
