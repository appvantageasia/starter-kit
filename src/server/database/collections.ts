import { Collection, Db } from 'mongodb';
import * as documents from './documents';

export type Collections = {
    users: Collection<documents.User>;
    topics: Collection<documents.Topic>;
};

export const getCollections = (db: Db): Collections => ({
    users: db.collection<documents.User>('users'),
    topics: db.collection<documents.Topic>('topics'),
});
