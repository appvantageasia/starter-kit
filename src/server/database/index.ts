export * from './documents';
export * from './collections';
export * from './instance';
export * from './migrate';
export * from './helpers';
