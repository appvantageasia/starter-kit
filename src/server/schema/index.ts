import { makeExecutableSchema } from 'graphql-tools';
import resolvers from './resolvers';
import typeDefs from './typeDefs.graphql';

const schema = makeExecutableSchema({ typeDefs: [typeDefs], resolvers });

export default schema;
