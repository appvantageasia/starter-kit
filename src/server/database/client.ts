import { MongoClient, MongoClientOptions } from 'mongodb';
import config from '../core/config';
import { getEncryptionKeyId } from './helpers';
import getKMS from './kms';
import buildSchemaMap from './schemaMap';

export const getRegularOptions = (): MongoClientOptions => ({
    maxPoolSize: config.db.pool,
});

export const getRegularClient = () => MongoClient.connect(config.db.uri, getRegularOptions());

export const getEncryptedClient = async (kms: Exclude<ReturnType<typeof getKMS>, null>) => {
    const keyId = await getEncryptionKeyId();

    if (!keyId) {
        throw new Error('No encryption key ID have been provided for the schema');
    }

    return MongoClient.connect(config.db.uri, {
        ...getRegularOptions(),

        // enable command monitoring for the client
        monitorCommands: true,

        autoEncryption: {
            keyVaultNamespace: kms.keyVaultNamespace,
            kmsProviders: kms.kmsProviders,
            schemaMap: buildSchemaMap(keyId),

            // configure cryptd client
            extraOptions: {
                mongocryptdURI: config.db.cryptd.uri,
                mongocryptdBypassSpawn: config.db.cryptd.mongocryptdBypassSpawn,
                mongocryptdSpawnArgs: config.db.cryptd.mongocryptdSpawnArgs,
            },
        },
    });
};
