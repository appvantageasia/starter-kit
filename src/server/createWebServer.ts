// create http server
import http, { IncomingMessage, Server } from 'http';
import { ApolloServer } from 'apollo-server-express';
import compression from 'compression';
import express, { Express } from 'express';
import { graphqlUploadExpress } from 'graphql-upload';
import schema from './schema';
import createContext, { Context, RootDocument } from './schema/context';

export type WebServerCreation = {
    httpServer: Server;
    expressServer: Express;
    apolloServer: ApolloServer;
};

const createWebServer = (): WebServerCreation => {
    const apolloServer = new ApolloServer({
        schema,
        // do not use the build-in upload types
        uploads: false,

        // enable tracing for development purposes
        tracing: process.env.NODE_ENV === 'development',

        // provide a custom context
        context: ({ req }: { req: IncomingMessage }): Promise<Context> => createContext(req),

        // provide a custom root document
        rootValue: (): RootDocument => null,
    });

    const expressServer = express();
    expressServer.disable('x-powered-by');
    expressServer.use(compression());
    expressServer.use(express.json());
    expressServer.use(express.urlencoded({ extended: true }));

    expressServer.use(
        '/graphql',
        graphqlUploadExpress(),
        apolloServer.getMiddleware({ bodyParserConfig: { limit: '50mb' }, path: '/' })
    );

    expressServer.use((req, res) => {
        res.send('hello world');
    });

    // then here comes our error handler
    // eslint-disable-next-line no-unused-vars
    expressServer.use((error, request, response, next) => {
        console.error(error);
        response.status('500').send('Internal error');
    });

    const httpServer = http.createServer(expressServer);

    return { httpServer, apolloServer, expressServer };
};

export default createWebServer;
