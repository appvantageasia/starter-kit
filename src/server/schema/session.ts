import { IncomingMessage } from 'http';
import { AuthenticationError } from 'apollo-server';
import cookie from 'cookie';
import jwt from 'jsonwebtoken';
import { getOr } from 'lodash/fp';
import { ObjectId } from 'mongodb';
import config from '../core/config';
import { generateToken, consumeToken, csrfGenerator } from '../core/tokens';
import { getDatabaseContext } from '../database';
import getIp from '../utils/getIp';

export type SessionData = {
    userId: ObjectId;
    sessionId: ObjectId;
};

export const getSessionToken = async (data: SessionData) => {
    const csrf = csrfGenerator();
    const token = generateToken('session', data, config.session.lifetime, csrf);

    // get iat and exp from token instead
    const { iat, exp } = jwt.decode(token, { json: true });

    return {
        token: generateToken('session', data, config.session.lifetime, csrf),
        csrf,
        exp: new Date(exp * 1000),
        iat: new Date(iat * 1000),
    };
};

export const readSessionToken = (token: string, csrf: string) => consumeToken<SessionData>('session', token, csrf);

export const getCSRFFromHeaders = (req: IncomingMessage): string => {
    const requestCookie = req.headers?.cookie;

    return requestCookie ? getOr('', 'CSRF', cookie.parse(requestCookie)) : '';
};

const getAuthorizationFromHeader = (req: IncomingMessage) => {
    const header = req.headers?.authorization as string;

    if (!header) {
        return null;
    }

    const [type, token] = header.split(' ');

    if (type !== 'Bearer') {
        throw new AuthenticationError('Unsupported authentication method');
    }

    if (!token) {
        throw new AuthenticationError('Missing token');
    }

    return token;
};

export const getSessionDataFromRequest = async (
    req: IncomingMessage,
    authorization?: string
): Promise<SessionData | null> => {
    const token = authorization || getAuthorizationFromHeader(req);

    if (!token) {
        return null;
    }

    const data = await (async () => {
        try {
            const csrf = getCSRFFromHeaders(req);

            return await readSessionToken(token, csrf);
        } catch (error) {
            throw new AuthenticationError('Invalid authentication token');
        }
    })();

    if (!data) {
        return null;
    }

    const { collections } = await getDatabaseContext();
    const result = await collections.userSessions.updateOne(
        { _id: data.sessionId },
        {
            $set: {
                // update user agent to keep the latest
                userAgent: req.headers['user-agent'],
                ip: getIp(req),
                // then update last activity date
                lastActivityAt: new Date(),
            },
        }
    );

    if (!result.matchedCount) {
        // session was invalidated
        throw new AuthenticationError('Session invalidated');
    }

    return data;
};
