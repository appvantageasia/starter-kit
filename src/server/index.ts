import chalk from 'chalk';
import { runValidityChecks } from './config';
import createWebServer from './createWebServer';
import { getDatabaseContext, migrate } from './database';
import { setup as startWorker } from './queues';

runValidityChecks();

const executeDataMigration = async (): Promise<void> => {
    // get the database connection
    const { db } = await getDatabaseContext();

    try {
        // then migrate the database
        await migrate(db);
    } catch (error) {
        // print it
        console.error(error);
        // and exit on status 1
        process.exit(1);
    }

    console.info(chalk.greenBright('Migration completed'));

    // and properly exit
    process.exit(0);
};

export { createWebServer, startWorker, executeDataMigration };
