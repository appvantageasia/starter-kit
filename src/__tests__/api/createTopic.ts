import path from 'path';
import { gql, ApolloError } from '@apollo/client';
import { ObjectId } from 'mongodb';
import { getDatabaseContext } from '../../server/database';
import { getSessionToken, SessionData } from '../../server/schema/session';
import {
    composeHandlers,
    loadFixtures,
    setupDatabase,
    cleanDatabase,
    setupWebService,
    getApolloClient,
    setupEmptyBucket,
    createBlobFrom,
} from '../helpers';
import fixtures from './authenticate.fixture.json';

const mutation = gql`
    mutation test($title: String!, $body: String!, $files: [Upload!]) {
        createTopic(title: $title, body: $body, attachments: $files) {
            id
        }
    }
`;

const webService = setupWebService();

beforeEach(composeHandlers(setupEmptyBucket, setupDatabase, loadFixtures(fixtures), webService.initialize));

afterEach(composeHandlers(cleanDatabase, webService.cleanUp));

test('Create topic requires authorizations', async () => {
    const client = getApolloClient(webService.url);
    const variables = { title: 'test', body: 'this is a test' };
    const promise = client.mutate({ mutation, variables });
    await expect(promise).rejects.toBeInstanceOf(ApolloError);
    const error: ApolloError = await promise.catch(error => error);
    expect(error.graphQLErrors).toMatchSnapshot();
});

test('Create topic work successfully without attachments', async () => {
    const originalData: SessionData = { userId: new ObjectId('601f7ae763d6cfc2554f6b67') };
    const originalToken = await getSessionToken(originalData);
    const client = getApolloClient(webService.url, { authorizationToken: originalToken });
    const variables = { title: 'test', body: 'this is a test' };
    const { data } = await client.mutate({ mutation, variables });
    expect(data.createTopic).not.toBeNull();
    expect(data.createTopic.id).not.toBeNull();
    const { collections } = await getDatabaseContext();
    const topic = await collections.topics.findOne({ _id: new ObjectId(data.createTopic.id) });
    expect(topic).not.toBeNull();
    expect(topic.title).toEqual(variables.title);
    expect(topic.body).toEqual(variables.body);
    expect(topic.attachments).toHaveLength(0);
    expect(topic.authorId.toHexString()).toEqual(originalData.userId.toHexString());
});

test('Create topic work successfully with attachments', async () => {
    const originalData: SessionData = { userId: new ObjectId('601f7ae763d6cfc2554f6b67') };
    const originalToken = await getSessionToken(originalData);
    const client = getApolloClient(webService.url, { authorizationToken: originalToken });
    const originalFile = await createBlobFrom(path.resolve(__dirname, './img.png'));
    const variables = { title: 'test', body: 'this is a test', files: [originalFile] };
    const { data } = await client.mutate({ mutation, variables });
    expect(data.createTopic).not.toBeNull();
    expect(data.createTopic.id).not.toBeNull();
    const { collections } = await getDatabaseContext();
    const topic = await collections.topics.findOne({ _id: new ObjectId(data.createTopic.id) });
    expect(topic).not.toBeNull();
    expect(topic.title).toEqual(variables.title);
    expect(topic.body).toEqual(variables.body);
    expect(topic.attachments).toHaveLength(1);
    expect(topic.authorId.toHexString()).toEqual(originalData.userId.toHexString());
});
