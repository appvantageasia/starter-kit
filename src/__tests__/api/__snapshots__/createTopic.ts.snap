// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create topic requires authorizations 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "UNAUTHENTICATED",
    },
    "locations": Array [
      Object {
        "column": 3,
        "line": 2,
      },
    ],
    "message": "Not authenticated",
    "path": Array [
      "createTopic",
    ],
  },
]
`;
